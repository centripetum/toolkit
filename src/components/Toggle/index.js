import * as React from 'react'

import { identity, not } from 'ramda'
import { isNilOrEmpty } from 'ramda-adjunct'
import styled from 'styled-components'

const { useEffect, useState } = React

const Label = styled.label`
  padding: 0.2rem;
  font-size: 1rem;
  font-family: sans-serif;
  margin: 0 0.2rem;
`
Label.displayName = 'ToggleLabel'

const Button = styled.button`
  border: 1px solid gray;
  border-color: ${({ showing }) => (showing ? 'green' : 'red')};
  border-radius: 0.2rem;
  padding: 0.2rem;
  font-size: 1rem;
  font-family: sans-serif;
  margin: 0 0.2rem;
  color: ${({ showing }) => (showing ? 'green' : 'red')};
`
Button.displayName = 'Toggle'

function getLabel (id, label) {
  if (isNilOrEmpty(label)) {
    return null
  }

  return <Label htmlFor={id}>{label}</Label>
}

export default function Toggle ({
  defaultValue = false,
  id,
  label,
  name,
  onChange,
  validate = identity
}) {
  const [value, setValue] = useState(defaultValue)

  const toggle = () => setValue(not(value))

  useEffect(() => onChange(name, value, validate(value)), [value])

  return (
    <div onClick={toggle}>
      {getLabel(id, label)}
      <Button id={id} name={name}>
        {value ? 'Yes' : 'No'}
      </Button>
    </div>
  )
}
