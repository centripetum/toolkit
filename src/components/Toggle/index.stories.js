import * as React from 'react'
import { storiesOf } from '@storybook/react'

import Toggle from './'

const id = 'Toggle'
const name = 'Toggle'
const label = 'Toggle'

storiesOf('components/Toggle', module)
  .add('Basic component', () => (
    <Toggle id={id} name={name} label={label} onChange={console.log} />
  ))
  .add('Custom button text', () => (
    <Toggle
      id={id}
      name={name}
      label={label}
      onChange={console.log}
      onText={'Selected'}
      offText={'Unselected'}
    />
  ))
  .add('Button images', () => (
    <Toggle
      id={id}
      name={name}
      label={label}
      onChange={console.log}
      onText={''}
      offText={''}
      onImage={'https://img.icons8.com/metro/26/000000/toggle-on.png'}
      offImage={'https://img.icons8.com/metro/26/000000/toggle-off.png'}
    />
  ))

//  kenobi
//  griveous
