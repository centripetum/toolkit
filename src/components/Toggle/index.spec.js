import * as React from 'react'
import { render } from 'react-testing-library'

import Toggle from './'

describe('components:Toggle', () => {
  it('Toggle is being created properly with default False value', () => {
    const { container } = render(
      <Toggle
        id='Toggle'
        name='Toggle'
        label='Toggle'
        onChange={() => undefined}
      />
    )

    expect(container).toMatchSnapshot()
  }),
    it('Toggle is being created properly with initial True value', () => {
      const { container } = render(
        <Toggle
          defaultValue
          id='Toggle'
          name='Toggle'
          label='Toggle'
          onChange={() => undefined}
        />
      )

      expect(container).toMatchSnapshot()
    }),
    it('Toggle is being created properly with no label', () => {
      const { container } = render(
        <Toggle
          id='Toggle'
          name='Toggle'
          label={null}
          onChange={() => undefined}
        />
      )

      expect(container).toMatchSnapshot()
    })
})
