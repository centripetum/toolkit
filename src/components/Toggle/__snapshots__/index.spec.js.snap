// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components:Toggle Toggle is being created properly with default False value 1`] = `
<div>
  <div>
    <label
      class="sc-bdVaJa dMsKTk"
      for="Toggle"
    >
      Toggle
    </label>
    <button
      class="sc-bwzfXH dRdGGI"
      id="Toggle"
      name="Toggle"
    >
      No
    </button>
  </div>
</div>
`;

exports[`components:Toggle Toggle is being created properly with initial True value 1`] = `
<div>
  <div>
    <label
      class="sc-bdVaJa dMsKTk"
      for="Toggle"
    >
      Toggle
    </label>
    <button
      class="sc-bwzfXH dRdGGI"
      id="Toggle"
      name="Toggle"
    >
      Yes
    </button>
  </div>
</div>
`;

exports[`components:Toggle Toggle is being created properly with no label 1`] = `
<div>
  <div>
    <button
      class="sc-bwzfXH dRdGGI"
      id="Toggle"
      name="Toggle"
    >
      No
    </button>
  </div>
</div>
`;
